### PR TITLE
RATIS-2049 Ensure atomic safety for volatile object

### PR DESCRIPTION
**Pull request by OneSizeFitsQuorum**
see [jira](https://issues.apache.org/jira/browse/RATIS-2049)

### CODE REVIEW

**Comment by OneSizeFitsQuorum** [comment]
@szetszwo PTAL

File: ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java [hunk 1]
```diff
@@ -157,7 +158,7 @@ synchronized int process(Event event) {
   private final int maxOutstandingInstallSnapshots;
   private final TimeoutExecutor scheduler = TimeoutExecutor.getInstance();
 
-  private volatile StreamObservers appendLogRequestObserver;
+  private AtomicReference<StreamObservers> appendLogRequestObserver = new AtomicReference<>();
```
Comment (szetszwo) [inline ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java:161]: Let's suppress the warning?  WRITE is protected by `writeLock` while READ is protected `volatile`.

File: ratis-common/src/main/java/org/apache/ratis/util/MemoizedCheckedSupplier.java [hunk 2]
```diff
@@ -44,6 +44,8 @@ public static <RETURN, THROW extends Throwable> MemoizedCheckedSupplier<RETURN,
   }
 
   private final CheckedSupplier<RETURN, THROW> initializer;
+
+  @SuppressWarnings({"squid:S3077"}) // Suppress volatile for immutable object warning
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/util/MemoizedCheckedSupplier.java:48]: Let's change the comment to
```java
// Suppress volatile for generic type
```

File: ratis-common/src/main/java/org/apache/ratis/util/MemoizedSupplier.java [hunk 3]
```diff
@@ -40,6 +40,7 @@ public static <T> MemoizedSupplier<T> valueOf(Supplier<T> supplier) {
   }
 
   private final Supplier<T> initializer;
+  @SuppressWarnings({"squid:S3077"}) // Suppress volatile for immutable object warning
```
Comment (szetszwo) [inline ratis-common/src/main/java/org/apache/ratis/util/MemoizedSupplier.java:43]: Let's change the comment to
```java
// Suppress volatile for generic type
```

**Review by szetszwo — COMMENTED** [review]
@OneSizeFitsQuorum , thanks for working on this!  Please see the comments inlined.

File: ratis-common/src/main/java/org/apache/ratis/util/MemoizedCheckedSupplier.java (hunk 2 above)
Comment (OneSizeFitsQuorum) [reply ratis-common/src/main/java/org/apache/ratis/util/MemoizedCheckedSupplier.java:48]: Fixed

File: ratis-common/src/main/java/org/apache/ratis/util/MemoizedSupplier.java (hunk 3 above)
Comment (OneSizeFitsQuorum) [reply ratis-common/src/main/java/org/apache/ratis/util/MemoizedSupplier.java:43]: Fixed

File: ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java (hunk 1 above)
Comment (OneSizeFitsQuorum) [reply ratis-grpc/src/main/java/org/apache/ratis/grpc/server/GrpcLogAppender.java]: Brilliant design! Changed!

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**